File: 0001-0100/0092-Reverse Linked List II/0092-Reverse Linked List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode dummy(-1);
        dummy.next = head;
        ListNode* p = &dummy;
        int i = 0;
        while (i < m - 1) {
            p = p->next;
            ++i;
        }
        
        ListNode* curr = p->next;
        ListNode* prev = nullptr;
        while (i < n) {
            ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
            ++i;
        }
        p->next->next = curr;
        p->next = prev;
        
        return dummy.next;
    }
};

File: 0001-0100/0092-Reverse Linked List II/0092-Reverse Linked List II.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        p = dummy = ListNode(-1)
        p.next = head
        for _ in range(m - 1):
            p = p.next
        curr = p.next
        prev = None
        for _ in range(n - m + 1):
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        p.next.next = curr
        p.next = prev
        return dummy.next
